exciting_environments/gym_wrapper.py

```diff
@@ -37,55 +75,111 @@ def step(self, action):
+            state, reward, self.env.env_properties
+        )
+        truncated = jax.vmap(self.generate_truncated, in_axes=(0, self.env.in_axes_env_properties))(
+            state, self.env.env_properties
+        )
         # transform dataclass to array
-        state = jnp.array(tree_flatten(state)[0]).T
+        state = tree_flatten(state)[0]  # jnp.array(tree_flatten(state)[0]).T
 
-        return obs, reward, terminated, truncated, state
+        return obs, reward, terminated, truncated, state, reference_hold_steps
 
     def reset(self, rng: chex.PRNGKey = None, initial_state: jdc.pytree_dataclass = None):
         """Resets environment to default or passed initial state."""
         # TODO: rng
 
         if initial_state is not None:
-            assert (
-                jnp.array(tree_flatten(self.env.init_state())[0]).T.shape == initial_state.shape
-            ), f"initial_state should have shape={jnp.array(tree_flatten(self.env.init_state())[0]).T.shape}"
-            obs, state = self.env.reset(initial_state=tree_unflatten(self.state_tree_struct, initial_state.T))
+            try:
+                _, _ = self.env.reset(initial_state=tree_unflatten(self.state_tree_struct, initial_state))
+            except:
+                print("initial_state should have the same structure as tree_flatten(self.env.vmap_init_state()")
+            obs, state = self.env.reset(initial_state=tree_unflatten(self.state_tree_struct, initial_state))
         else:
-            obs, state = self.env.reset()
-        self.state = jnp.array(tree_flatten(state)[0]).T
+            if rng is not None:
+                obs, state = self.env.reset(rng)
+            else:
+                PRNGKey = jax.vmap(jax.random.PRNGKey)(np.random.randint(0, 2**31, size=(self.env.batch_size,)))
+                obs, state = self.env.reset(PRNGKey)
+
+        state, self.reference_hold_steps = jax.vmap(
+            self.generate_new_ref, in_axes=(0, self.env.in_axes_env_properties, 0)
+        )(state, self.env.env_properties, jnp.zeros(self.env.batch_size))
+
+        self.state = tree_flatten(state)[0]  # jnp.array(tree_flatten(state)[0]).T
         return obs, {}
 
+    def update_ref(self, state, env_properties, hold_steps):
+        state, hold_steps = jax.lax.cond(
+            hold_steps[0] == 0, self.generate_new_ref, lambda a, b, c: (a, c), state, env_properties, hold_steps
+        )
+        hold_steps += -1
+        return state, hold_steps
+
+    def generate_new_ref(self, state, env_properties, hold_steps):
+        with jdc.copy_and_mutate(state, validate=False) as new_state:
+            init = self.env.init_state(env_properties, state.PRNGKey)
+            for name in self.control_state:
+                setattr(new_state.reference, name, getattr(init.physical_state, name))
+
+            hold_steps = jax.random.randint(
+                init.PRNGKey,
+                minval=self.ref_params["hold_steps_min"],
```

Comment:
i think i prefer these to be passed to the function and not be used through `self.` like this. Also is it a problem that this is a dict? Can this lead to inconsistent behaviour? Can you change this between resets?